File: QM/QM.pl
use strict;
use warnings;

# Initialize global variables
my @product_terms = ();  # Store all the minterms
my @essential_implicants = ();  # Store essential implicants

# Create the Prime Implicant Chart mapping minterms to prime implicants
sub create_prime_implicant_chart {
    my ($minterms_ref, $prime_implicants_ref) = @_;
    my %chart = ();  # Chart to store minterms and associated implicants

    # Check which prime implicants cover each minterm
    for my $minterm (@$minterms_ref) {
        for my $implicant (@$prime_implicants_ref) {
            if (covers($minterm, $implicant)) {
                push @{$chart{$minterm}}, $implicant;
            }
        }
    }
    
    return \%chart;
}

# Function to check if a prime implicant covers a minterm
sub covers {
    my ($minterm, $implicant) = @_;
    
    # Compare each bit of the minterm and implicant
    for my $i (0 .. length($minterm) - 1) {
        if (substr($implicant, $i, 1) ne '-' && substr($minterm, $i, 1) ne substr($implicant, $i, 1)) {
            return 0;  # False if the bits differ
        }
    }
    return 1;  # True if it covers
}

# Select essential prime implicants from the chart
sub select_essential_prime_implicants {
    my ($prime_implicant_chart_ref) = @_;
    my %covered_minterms = ();  # Keep track of covered minterms
    my @essential_implicants = ();  # List of essential implicants

    # Go through each minterm in the chart
    while (my ($minterm, $implicants_ref) = each %$prime_implicant_chart_ref) {
        if (@$implicants_ref == 1) {  # If only one implicant covers the minterm
            push @essential_implicants, $implicants_ref->[0];
            $covered_minterms{$minterm} = 1;
        }
    }

    # Remove covered minterms from the chart
    for my $minterm (keys %covered_minterms) {
        delete $prime_implicant_chart_ref->{$minterm};
    }

    # Return essential implicants and remaining chart
    return (\@essential_implicants, $prime_implicant_chart_ref);
}

# Combine two terms if they differ by exactly one bit
sub combine_terms {
    my ($term1, $term2) = @_;
    my $diff_count = 0;
    my @combined = ();

    for my $i (0 .. length($term1) - 1) {
        if (substr($term1, $i, 1) ne substr($term2, $i, 1)) {
            $diff_count++;
            push @combined, '-';  # Replace differing bit with '-'
        } else {
            push @combined, substr($term1, $i, 1);
        }
    }

    return $diff_count == 1 ? join('', @combined) : undef;
}

# Group minterms by the number of 1s in their binary representation
sub group_minterms {
    my ($minterms_ref) = @_;
    my %groups = ();

    for my $term (@$minterms_ref) {
        my $num_ones = $term =~ tr/1//;
        push @{$groups{$num_ones}}, $term;
    }

    return \%groups;
}

# Combine all terms by iterating through adjacent groups
sub combine_all_terms {
    my ($minterms_ref) = @_;
    my @all_prime_implicants = ();
    my $grouped_terms_ref = group_minterms($minterms_ref);

    while (keys %$grouped_terms_ref) {
        my %next_grouped_terms = ();
        my %used_terms = ();

        for my $num_ones (sort keys %$grouped_terms_ref) {
            if ($grouped_terms_ref->{$num_ones + 1}) {
                for my $term1 (@{$grouped_terms_ref->{$num_ones}}) {
                    for my $term2 (@{$grouped_terms_ref->{$num_ones + 1}}) {
                        my $combined = combine_terms($term1, $term2);
                        if ($combined) {
                            push @{$next_grouped_terms{($combined =~ tr/1//)}}, $combined;
                            $used_terms{$term1} = 1;
                            $used_terms{$term2} = 1;
                        }
                    }
                }
            }
        }

        # Add unused terms as prime implicants
        push @all_prime_implicants, grep { !$used_terms{$_} } map { @{$grouped_terms_ref->{$_}} } keys %$grouped_terms_ref;
        $grouped_terms_ref = \%next_grouped_terms;
    }

    return \@all_prime_implicants;
}

# Output PLA format with essential prime implicants
sub output_pla_format {
    my ($inputs, $outputs, $essential_implicants_ref, $output_file) = @_;
    
    # Handle empty essential implicants array
    my $pla_output = ".i $inputs\n.o $outputs\n.p " . scalar(@$essential_implicants_ref) . "\n";
    if (@$essential_implicants_ref) {
        $pla_output .= join("\n", @$essential_implicants_ref) . "\n";
    } else {
        $pla_output .= "\n";  # Handle empty case
    }
    $pla_output .= ".e\n";

    print "\nPLA Format Output (Terminal):\n$pla_output";

    # Write to the output file
    open my $fh, '>', $output_file or die "Could not open '$output_file' for writing: $!\n";
    print $fh $pla_output;
    close $fh;

    print "\nPLA file saved to $output_file\n";
}

# Process the input PLA file
sub process_pla_file {
    my ($file_path, $output_file) = @_;
    my @product_terms = ();
    my ($inputs, $outputs);

    open my $fh, '<', $file_path or die "Could not open '$file_path' for reading: $!\n";

    while (my $line = <$fh>) {
        chomp $line;
        if ($line =~ /^\.i\s+(\d+)/) {
            $inputs = $1;
        } elsif ($line =~ /^\.o\s+(\d+)/) {
            $outputs = $1;
        } elsif ($line =~ /^\.p\s+(\d+)/) {
            # We can ignore this part for now
        } elsif ($line =~ /\|/) {
            push @product_terms, (split /\|/, $line)[0];
        }
    }

    close $fh;

    # Validate product terms
    if (@product_terms == 0) {
        die "No product terms found in the PLA file!\n";
    }

    # Process the product terms to find the essential prime implicants
    my @prime_implicants = @{combine_all_terms(\@product_terms)};
    my $prime_implicant_chart_ref = create_prime_implicant_chart(\@product_terms, \@prime_implicants);
    my ($essential_implicants_ref, $remaining_chart_ref) = select_essential_prime_implicants($prime_implicant_chart_ref);

    # Output PLA format and save to file
    output_pla_format($inputs, $outputs, $essential_implicants_ref, $output_file);
}

# Main execution
process_pla_file('example_9minterms_1output.pla', 'output2.pla');

File: QM/QM.py
# Bder Abuzayed
# EEE 4334
# Quine–McCluskey algorithm in python
product_terms = []    #Our global list that stores all the minterms
essential_implicants = [] #changed to global variable to pass it on the function to produce the output

def create_prime_implicant_chart(minterms, prime_implicants):
    """Create the Prime Implicant Chart mapping minterms to prime implicants."""
    chart = {minterm: [] for minterm in minterms}

    # Check which prime implicants cover each minterm
    for minterm in minterms:
        for implicant in prime_implicants:
            if covers(minterm, implicant):
                chart[minterm].append(implicant)

    # Display the chart in a readable format
    #print("\nPrime Implicant Chart:")
    #print(f"{'Minterms':<15} | {'Prime Implicants':<50}")
    #print("-" * 70)
    #for minterm, implicants in chart.items():
       # print(f"{minterm:<15} | {', '.join(implicants):<50}")

    return chart

# Function to check if a prime implicant covers a minterm
def covers(minterm, implicant):
    
    for m_bit, i_bit in zip(minterm, implicant):
        if i_bit != '-' and m_bit != i_bit:
            return False
    return True

def select_essential_prime_implicants(prime_implicant_chart):
    essential_implicants = set()
    covered_minterms = set()

    # Go through each minterm in the chart
    for minterm, implicants in prime_implicant_chart.items():
        if len(implicants) == 1:  # If only one implicant covers the minterm
            essential_implicant = implicants[0]
            if essential_implicant not in essential_implicants:
                essential_implicants.add(essential_implicant)
                
            # Mark the minterm as covered
            covered_minterms.add(minterm)
    
    # Remove covered minterms from the chart for next round
    for minterm in covered_minterms:
        del prime_implicant_chart[minterm]
    
    return essential_implicants, prime_implicant_chart

def combine_terms(term1, term2):
    """Combine two minterms that differ by exactly one bit."""
    diff_count = 0
    combined = []

    # Check that both terms have the same length
    if len(term1) != len(term2):
        print(f"Error: Terms have different lengths: {term1} and {term2}")
        return None

    for bit1, bit2 in zip(term1, term2):
        if bit1 != bit2:
            diff_count += 1
            combined.append('-')  # Replace differing bit with '-'
        else:
            combined.append(bit1)

    # If they differ by exactly one bit, return the combined term
    if diff_count == 1:
        #print(f"Combined {term1} and {term2} to {''.join(combined)}")  # Debug print
        return ''.join(combined)
    else:
        return None


def combine_all_terms(minterms):
    """Combine minterms that differ by one bit, repeat until no more combinations can be made."""
    all_prime_implicants = []  # List to store prime implicants
    grouped_terms = group_minterms(minterms)  # Group terms by the number of 1s
    
    while grouped_terms:
        next_grouped_terms = {}  # Prepare for the next iteration
        used_terms = set()       # Track terms that were combined in this iteration
        prime_candidates = set()  # Track terms that remain unused

        # Iterate through adjacent groups to combine terms
        for num_ones in sorted(grouped_terms.keys()):
            if num_ones + 1 in grouped_terms:  # Check for adjacent groups
                for term1 in grouped_terms[num_ones]:
                    for term2 in grouped_terms[num_ones + 1]:
                        combined = combine_terms(term1, term2)
                        if combined:
                            group_key = combined.count('1')  # Group based on new count of 1s
                            if group_key not in next_grouped_terms:
                                next_grouped_terms[group_key] = []
                            if combined not in next_grouped_terms[group_key]:
                                next_grouped_terms[group_key].append(combined)
                            used_terms.add(term1)
                            used_terms.add(term2)
            
            # Track terms that were not combined
            for term in grouped_terms[num_ones]:
                if term not in used_terms:
                    prime_candidates.add(term)

        # Add unused terms as prime implicants
        all_prime_implicants.extend(prime_candidates)
        
        # Prepare for the next iteration
        grouped_terms = remove_duplicates(next_grouped_terms)

    return all_prime_implicants

def remove_duplicates(grouped_terms):
    for key in grouped_terms:
        grouped_terms[key] = list(set(grouped_terms[key]))
    return grouped_terms

def group_minterms(minterms):
    """Group minterms by the number of 1s in their binary representation."""
    groups = {}

    for term in minterms:
        # Remove the part after | (if exists), and only count the binary part
        term = term.split('|')[0]
        num_ones = term.count('1')
        if num_ones not in groups:
            groups[num_ones] = []
        groups[num_ones].append(term)
    
    return groups

def output_pla_format(inputs, outputs, essential_implicants, output_file):
    # Create the header for the PLA file
    pla_output = f".i {inputs}\n.o {outputs}\n.p {len(essential_implicants)}\n"
    
    # Generate the product terms for the essential prime implicants
    for implicant in essential_implicants:
        pla_output += implicant + "\n"

    # Mark the end of the product terms
    pla_output += ".e\n"
    
    # Print to terminal 
    print("\nPLA Format Output (Terminal):")
    print(pla_output)

    # Write the PLA format file
    try:
        with open(output_file, 'w') as file:
            file.write(pla_output)
        print(f"\nPLA file saved to {output_file}")
    except Exception as e:
        print(f"Error saving PLA file: {str(e)}")

def process_pla_file(file_path, output_file):
    inputs = None
    outputs = None
    global product_terms

    try:
        with open(file_path, 'r') as file:
            for line in file:
                line = line.strip()  # Remove whitespace. THIS STEP IS VERY IMPORTANT becasue it will mess up the parsing if not done correctly
                
                if line.startswith('.i'):
                    inputs = int(line.split()[1])  # No. of inputs

                elif line.startswith('.o'):
                    outputs = int(line.split()[1])  # No. of outputs

                elif line.startswith('.p'):
                    product_term_count = int(line.split()[1])  # Store product terms count
                # Parse product terms themselves
                elif "|" in line:
                    product_terms.append(line.split('|')[0].strip())  # Store product term


        # Limit the product terms to the first 10 for debugging
        # Save the product terms (just for debugging)
        #print(f"First 5 Product Terms (After Parsing): {product_terms[:10]}")  
        
        # Call process_product_terms to generate the essential prime implicants
        process_product_terms()

        # Output the PLA format using the essential prime implicants and save to the file
        output_pla_format(inputs, outputs, essential_implicants, output_file)

        # Output the results
        print(f"Inputs: {inputs}")
        print(f"Outputs: {outputs}")
        print(f"Minterms: {product_term_count}")
        #print(f"Product Terms: {product_terms}")

    except FileNotFoundError:
        print(f"Error: The file '{file_path}' was not found.")
    except ValueError as ve:
        print(f"ValueError: {ve}. This typically happens if there's an issue with the data format in the file.")
    except Exception as e:
        print(f"An unexpected error occurred in process_pla_file: {str(e)}")

def process_product_terms():
    # Combine the product terms (we limit it to the first 5 for now)
    first_five_terms = product_terms[:] 
    #print(f"First 5 Product Terms: {first_five_terms}")  # Debugging line
    
    combined_terms = combine_all_terms(first_five_terms) # prime implicant generated and stored into combiner_terms
    
    # Output the results
    #print(f"Prime Implicants: {combined_terms}")

    # Create and display the Prime Implicant Chart
    prime_implicant_chart = create_prime_implicant_chart(first_five_terms, combined_terms)
    
    
    global essential_implicants  # Make essential_implicants accessible to the output function
    # Select essential prime implicants
    essential_implicants, remaining_chart = select_essential_prime_implicants(prime_implicant_chart)
    
    # Output the results
    print(f"Essential Prime Implicants: {essential_implicants}")
 
# The Code reads a static file that already exists. I will include the file in the Zip
process_pla_file('example_7inputs_99minterms.pla', 'output.pla')   # Calls the function to parse the input from the pla format
process_product_terms()  # Now call the function to combine the product terms